### ISSUE TITLE
An alternative @overload syntax for JSDoc

### ISSUE THREAD

**Issue by rdking** [microsoft/TypeScript#52796]
# Suggestion

## 🔍 Search Terms

jsdoc
function
overload

## ✅ Viability Checklist
My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Here's an alternative thought on the syntax that could satisfy [PR #52474](https://github.com/microsoft/TypeScript/pull/52474) without causing parser issues mentioned in [PR #52577](https://github.com/microsoft/TypeScript/pull/52577). What if overload simply allowed you to declare a method signature? For instance:

```js
/**
 * @overload { (inst, klass, members) => object }
 * @overload { (inst, members) => undefined }
 * @param {object} inst
 * @param {function} klass
 * @param {object} members
 * @returns {object|undefined}
 */
function share(inst, klass, members) {}
```

This should be both easy to parse, and satisfy the user's needs while allowing for minimally redundant documentation, all in a single comment.

**Comment by fatcerberus** [comment]
I was under the impression that the *main purpose* of `@overload` was to allow you to have separate documentation (summary text, parameter descriptions, etc.) for each overload, like you can already do in TS.  Consolidating them like this would defeat the purpose.

**Comment by rdking** [comment]
Not true, or at least not with the way I envision it functioning when generating documentation. The presence of 2 or more `@overload` tags would have the effect of producing a function definition for each `@overload` supplied. The params given in the `@overload` tag would be used as a filter in the list of params within the jsdoc comment. The net effect of the above would be documented as if the example below didn't cause the latter to replace the former:

```js
/**
 * @function share
 * @param {object} inst
 * @param {function} klass
 * @param {object} members
 * @returns {object}
 */
/**
 * @function share
 * @param {object} inst
 * @param {object} members
 * @returns {undefined}
 */
```

**Comment by xiaoxiangmoe** [comment]
Temp solution ：

```ts
function plus1(n: number): number;
function plus1(n: bigint): bigint;
function plus1(n: number | bigint): number | bigint {
    return typeof s === 'bigint' ? s + 1n : s + 1;
}
```

```js
const plus1 = /** @type {{ (n: number): number, (n: bigint): bigint }}*/ (
    function plus1(/** @type {number | bigint}*/ s) {
        return typeof s === 'bigint' ? s + 1n : s + 1;
    }
)
```
